mlir/include/mlir/Support/SystemDesc.h

```diff
@@ -0,0 +1,514 @@
+    if (json_value.length() > 0 && json_value[0] == '[') {
+      // Parse as an array
+      llvm::Expected<std::vector<int64_t>> ivv =
+          llvm::json::parse<std::vector<int64_t>>(json_value);
+      if (ivv) {
+        *this = this->setProperty(name, ivv.get());
+        return *this;
+      }
+
+      llvm::Expected<std::vector<double>> idv =
+          llvm::json::parse<std::vector<double>>(json_value);
+      if (idv) {
+        *this = this->setProperty(name, idv.get());
+        return *this;
+      }
+    } else {
+      // int64_t because llvm::json has int64_t support (not int)
+      llvm::Expected<int64_t> iv = llvm::json::parse<int64_t>(json_value);
+      if (iv) {
+        *this = this->setProperty(name, iv.get());
+        return *this;
+      }
+
+      // Int type failed, try float now.
+      // double because llvm::json has double support (not float)
+      llvm::Expected<double> dv = llvm::json::parse<double>(json_value);
+      if (dv) {
+        *this = this->setProperty(name, dv.get());
+        return *this;
+      }
+    }
+
+    llvm::report_fatal_error(
+        "Neither int/float/vector value in Device Description: key " + name);
+  }
+
+  /// Get ID
+  DeviceID getID() const { return ID; }
+  /// Get device type
+  DeviceType getType() const { return type; }
+  /// Get device description
+  std::string getDescription() const { return description; }
+  /// Get all of device properties
+  const DevicePropertiesMapTy &getProperties() const {
+    return deviceProperties;
+  }
+  /// Get property value: returns the value of the property with given name, if
+  /// it exists. Otherwise throws exception (TODO)
+  std::optional<int64_t> getPropertyValueAsInt(llvm::StringRef name) const {
+    // check that property with the given name exists
+    auto iter = deviceProperties.find(std::string(name));
+    if (iter == deviceProperties.end()) {
+      return std::nullopt;
+    }
+    // TODO: we can do a tag check here.
+    return iter->second.data.iValue;
+  }
+  std::optional<double> getPropertyValueAsFloat(llvm::StringRef name) const {
+    // check that property with the given name exists
+    auto iter = deviceProperties.find(std::string(name));
+    if (iter == deviceProperties.end()) {
+      return std::nullopt;
+    }
+    // TODO: we can do a tag check here.
+    return iter->second.data.dValue;
+  }
+
+  /// Special functions
+  auto getAllDevicePropertyNames() const {
+    return llvm::map_range(
+        deviceProperties,
+        [](const DevicePropertiesMapTy::value_type &item) -> llvm::StringRef {
+          return item.first;
+        });
+  }
+
+  /// We use a list of key-value pairs to represent a system description in
+  /// JSON.
+  using DeviceDescJSONTy = std::map<std::string, std::string>;
+  static DeviceDesc
+  parseDeviceDescFromJSON(const DeviceDescJSONTy &device_desc);
+
+  // -----------------------------------------------------------------------
+  //          CPU specific methods
```

Comment:
IMHO specific things like this shouldn't be in the core DeviceDesc.

Reply:
Thanks for the comment @mikeurbach. Moved this out of the core class now. Pls take a look.